ossdbtoolsservice/metadata/contracts/schema_metadata.py

```diff
@@ -0,0 +1,146 @@
+        table_name
+        \tcolumn_name/type
+        \tcolumn_name/type
+
+        """)
+
+        self.cur.execute(
+            f"""
+            SELECT
+                t.table_name,
+                column_name,
+                data_type
+            FROM
+                information_schema.tables AS t
+                INNER JOIN information_schema.columns AS c
+                    ON t.table_name = c.table_name
+            WHERE
+                t.table_schema = '{self.schema}'
+            AND t.table_name NOT LIKE '%\_p%' --exclude partitions
+            AND t.table_name NOT LIKE 'pg\_%' --exclude system tables
+            ORDER BY
+                t.table_name,
+                c.ordinal_position;
+            """
+        )
+        tables = self.cur.fetchall()
+
+        cur_table = None
+        for table, column, dt in tables:
+            if table != cur_table:
+                cur_table = table
+                results += table + "\n"
+            results += f"\t{column}/{_pretty_type(dt)}\n"
+
+        return results
+
+    def describe_constraints(self) -> str:
+        results = dedent("""
+        ## Table constraints, in the form:
+        table_name
+        \tconstraint_def
+        \tconstraint_def
+
+        """)
+
+        self.cur.execute(
+            f"""
+            SELECT
+                t.relname AS table_name,
+                pg_get_constraintdef(c.oid) AS definition
+            FROM pg_constraint c
+            JOIN pg_class t ON c.conrelid = t.oid
+            WHERE c.contype IN ('p', 'u', 'f')
+                AND t.relname IN (
+                    SELECT tablename
+                    FROM pg_catalog.pg_tables
+                    WHERE schemaname = '{self.schema}'
+                    AND tablename NOT LIKE '%\_p%' --exclude partitions
+                    AND tablename NOT LIKE 'pg\_%' --exclude system tables
+                )
+            ORDER BY 1, 2;
+            """
+        )
+        tables = self.cur.fetchall()
+
+        cur_table = None
+        for table, constraint in tables:
+            if table != cur_table:
+                cur_table = table
+                results += table + "\n"
+            results += "\t" + _pretty_constraint(constraint) + "\n"
+
+        return results
+
+    def describe_indexes(self) -> str:
+        results = dedent("""
+        ## Table indexes, in the form:
+        table_name
+        \ttype (column_name, column_name)
+        \ttype (column_name, column_name)
+
+        """)
+
+        self.cur.execute(
+            f"""
+            SELECT tablename, indexdef
+            FROM pg_indexes
+            WHERE schemaname = '{self.schema}'
+            AND tablename NOT LIKE '%\_p%' --exclude partitions
+            AND tablename NOT LIKE 'pg\_%' --exclude system tables
+            ORDER BY 1, 2;
+            """
+        )
+        tables = self.cur.fetchall()
+        cur_table = None
+        for table, indx in tables:
```

Comment:
It's less likely to be performance issue here but in case of large schema, it's always better to use list append + join than string concatenation.

Reply:
I gave that a try in https://github.com/microsoft/pgtoolsservice/pull/459/commits/dc8d5d9b0c6cd8283556342996d35008cd75b0fe -- see what you think. I'm not sure whether that particular implementation really will go faster with all the list comprehensions and dictionary lookups, but curious what your opinion is.